controllers/csi/provisioner/agent.go

```diff
@@ -57,85 +57,75 @@ func installAgent(installAgentCfg *installAgentConfig) error {
 	}
 	logger.Info("Saved OneAgent package", "dest", tmpFile.Name())
 
-	fileInfo, err := tmpFile.Stat()
-	if err != nil {
-		return fmt.Errorf("failed to determine agent archive file size: %w", err)
-	}
-
-	zipr, err := zip.NewReader(tmpFile, fileInfo.Size())
-	if err != nil {
-		return fmt.Errorf("failed to open ZIP file: %w", err)
-	}
-
 	logger.Info("Unzipping OneAgent package")
-	if err := unzip(zipr, installAgentCfg); err != nil {
+	if err := unzip(tmpFile, installAgentCfg); err != nil {
 		return fmt.Errorf("failed to unzip file: %w", err)
 	}
-
 	logger.Info("Unzipped OneAgent package")
 
 	return nil
 }
 
-func unzip(r *zip.Reader, installAgentCfg *installAgentConfig) error {
-	outDir := installAgentCfg.targetDir
-	logger := installAgentCfg.logger
+func unzip(file afero.File, installAgentCfg *installAgentConfig) error {
+	target := installAgentCfg.targetDir
 	fs := installAgentCfg.fs
 
-	_ = fs.MkdirAll(outDir, 0755)
+	if file == nil {
+		return fmt.Errorf("file is nil")
+	}
 
-	// Closure to address file descriptors issue with all the deferred .Close() methods
-	extract := func(zipf *zip.File) error {
-		rc, err := zipf.Open()
-		if err != nil {
-			return err
-		}
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return fmt.Errorf("unable to determine file info: %w", err)
+	}
 
-		defer func() {
-			if err := rc.Close(); err != nil {
-				logger.Error(err, "Failed to close ZIP entry file", "path", zipf.Name)
-			}
-		}()
+	reader, err := zip.NewReader(file, fileInfo.Size())
+	if err != nil {
+		return fmt.Errorf("failed to open ZIP file: %w", err)
+	}
 
-		path := filepath.Join(outDir, zipf.Name)
+	_ = fs.MkdirAll(target, 0755)
 
-		// Check for ZipSlip (Directory traversal)
-		if !strings.HasPrefix(path, filepath.Clean(outDir)+string(os.PathSeparator)) {
-			return fmt.Errorf("illegal file path: %s", path)
-		}
+	for _, file := range reader.File {
+		err := func() error {
+			path := filepath.Join(target, file.Name)
 
-		mode := zipf.Mode()
+			// Check for ZipSlip: https://snyk.io/research/zip-slip-vulnerability
+			if !strings.HasPrefix(path, filepath.Clean(target)+string(os.PathSeparator)) {
+				return fmt.Errorf("illegal file path: %s", path)
+			}
 
-		// Mark all files inside ./agent/conf as group-writable
-		if zipf.Name != agentConfPath && strings.HasPrefix(zipf.Name, agentConfPath) {
-			mode |= 020
-		}
+			mode := file.Mode()
 
-		if zipf.FileInfo().IsDir() {
-			return fs.MkdirAll(path, mode)
-		}
+			// Mark all files inside ./agent/conf as group-writable
+			if file.Name != agentConfPath && strings.HasPrefix(file.Name, agentConfPath) {
+				mode |= 020
+			}
```

Comment:
Why do we need this?

Reply:
There is a [unit test](https://github.com/Dynatrace/dynatrace-operator/blob/master/controllers/csi/provisioner/agent_test.go#L201) that asserts this behavior, so I kept the same functionality.